showcase/src/elm/UrlGenerator.elm

```diff
@@ -0,0 +1,513 @@
+
+
+fromSourceCode : String -> String -> String
+fromSourceCode version elmCode =
+    let
+        { title, code } =
+            elliefy elmCode
+
+        titleUrl =
+            encodeUrl title
+
+        elmCodeUrl =
+            encodeUrl code
+
+        htmlCodeUrl =
+            encodeUrl htmlCode
+
+        -- Package parsing is interesting because every package has its own '&packages='
+        packagesUrl =
+            packages version
+                |> List.map (packageToString >> encodeUrl)
+                |> List.map (\p -> "&packages=" ++ p)
+                |> String.join ""
+
+        link =
+            "https://ellie-app.com/a/example/v1?title="
+                ++ titleUrl
+                ++ "&elmcode="
+                ++ elmCodeUrl
+                ++ "&htmlcode="
+                ++ htmlCodeUrl
+                ++ packagesUrl
+                ++ "&elmversion=0.19.0"
+    in
+    link
+
+
+
+---- HTML
+
+
+htmlCode : String
+htmlCode =
+    """<html>
+<head>
+  <style>
+    /* you can style your program here */
+  </style>
+</head>
+<body>
+  <main></main>
+  <script>
+    var app = Elm.Main.init({ node: document.querySelector('main') })
+    // you can use ports and stuff here
+  </script>
+</body>
+</html>
+"""
+
+
+
+---- PACKAGES
+
+
+type alias Package =
+    { name : String
+    , version : String
+    }
+
+
+
+-- Elm-antd will have dynamic versioning
+
+
+packages : String -> List Package
+packages version =
+    [ Package "elm/browser" "1.0.2"
+    , Package "elm/core" "1.0.5"
+    , Package "elm/html" "1.0.0"
+    , Package "supermacro/elm-antd" version
+    ]
+
+
+packageToString : Package -> String
+packageToString p =
+    p.name ++ "@" ++ p.version
+
+
+
+---- ENCODING
+-- replacing every reserved URL character using (most likely) the most inefficient way possible
+-- reserved characters (from Wikipedia): ! * ' ( ) ; : @ & = + $ , / ? % # [ ]
+
+
+encodeUrl : String -> String
+encodeUrl =
+    String.replace "%" "%25"
+        >> String.replace "\n" "%0A"
+        >> String.replace " " "%20"
+        >> String.replace "\"" "%22"
+        >> String.replace "!" "%21"
+        >> String.replace "#" "%23"
+        >> String.replace "$" "%24"
+        >> String.replace "&" "%26"
+        >> String.replace "'" "%27"
+        >> String.replace "(" "%28"
+        >> String.replace ")" "%29"
+        >> String.replace "*" "%2A"
+        >> String.replace "+" "%2B"
+        >> String.replace "," "%2C"
+        >> String.replace "/" "%2F"
+        >> String.replace ":" "%3A"
+        >> String.replace ";" "%3B"
+        >> String.replace "=" "%3D"
+        >> String.replace "?" "%3F"
+        >> String.replace "@" "%40"
+        >> String.replace "[" "%5B"
+        >> String.replace "]" "%5D"
+
+
+
+-------------------
+-- PARSING STUFF
+-- Elliefy makes the code renderable in Ellie. WIP
+-------------------
+
+
+elliefy : String -> { title : String, code : String }
```

Comment:
I definitely should have left a comment regarding this haha. Returning a `Result` would probably be an end goal, but I just wanted to have a way to quickly see the errors if something went wrong for debugging, since this isn't really production ready.

If the function returned a result, we would have to find a way to somehow view it in the HTML page, and that'll get a bit annoying. Maybe `Debug.log` would have been a better strategy though, but I think `Debug` functions break elm optimizations.

What do you think? If we returned a result, the `Container` module wouldn't really have a way to use the error, other than logging it. For production, and if we write enough tests for the function, maybe it wouldn't matter too much.

Reply:
I think that you could change the type signature of `elliefy` to return a result, and then you would pipe the output to `Result.withDefault`. 

```elm
        { title, code } =
            FileParser.elliefy elmCode
            |> Result.withDefault { title = "parse error", code = "..." }
```

That way the type signature can already be something more appropriate.

---

>  we would have to find a way to somehow view it in the HTML page

On Ellie you mean?

---

Re: usage of `Debug` isn't allowed for production builds if I'm not mistaken, but I'm ok committing `Debug.log` calls if this also works for you. I would actually prefer going with this approach rather than returning a non-sensical record. 

And once it's ready to deploy, we can remove the `Debug.log` calls.

Reply:
> On Ellie you mean?

I meant on the showcase, but you answered that question when you said "it's ok to make clicking on the ellie link icon a noop". 

<hr />

So I changed up the type signatures a bit more, and they should make a lot more sense.

As of now both `FileParser.elliefy` and `EllieLinks.fromSourceCode` both return `Result String String`, with the error type just being a stringified version of the parser `DeadEnds` type. Maybe we can define our own error type once the file parser gets more advanced? 

`EllieLinks.fromSourceCode` is currently the one that `Debug.log`s the error (temporary).

`Container.setSourceCode` takes care of this and transforms the `Result String String` to a `Maybe String` that's used in the component pages. This should mean any parsing errors would just result in the button being unclickable.